### PR TITLE
[SPARK-43009][SQL] Parameterized `sql()` with `Any` constants

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to change API of parameterized SQL, and replace type of argument values from `string` to `Any` in Scala/Java/Python and `Expression.Literal` in protobuf API. Language API can accept `Any` objects from which it is possible to construct literal expressions.

#### Scala/Java:

```scala
  def sql(sqlText: String, args: Map[String, Any]): DataFrame
```
values of the `args` map are wrapped by the `lit()` function which leaves `Column` as is and creates a literal from other Java/Scala objects (for more details see the `Scala` tab at https://spark.apache.org/docs/latest/sql-ref-datatypes.html).

#### Python:

```python
def sql(self, sqlQuery: str, args: Optional[Dict[str, Any]] = None, **kwargs: Any) -> DataFrame:
```
Similarly to Scala/Java `sql`, Python's `sql()` accepts Python objects as values of the `args` dictionary (see more details about acceptable Python objects at https://spark.apache.org/docs/latest/sql-ref-datatypes.html). `sql()` converts dictionary values to `Column` literal expressions by `lit()`.

#### Protobuf:

```proto
message SqlCommand {
  // (Required) SQL Query.
  string sql = 1;

  // (Optional) A map of parameter names to literal expressions.
  map<string, Expression.Literal> args = 2;
}
```

For example:
```scala
scala> val sqlText = """SELECT s FROM VALUES ('Jeff /*__*/ Green'), ('E\'Twaun Moore') AS t(s) WHERE s = :player_name"""
sqlText: String = SELECT s FROM VALUES ('Jeff /*__*/ Green'), ('E\'Twaun Moore') AS t(s) WHERE s = :player_name

scala> sql(sqlText, args = Map("player_name" -> lit("E'Twaun Moore"))).show(false)
+-------------+
|s            |
+-------------+
|E'Twaun Moore|
+-------------+
```

### Why are the changes needed?
The current implementation the parameterized `sql()` requires arguments as string values parsed to SQL literal expressions that causes the following issues:
1. SQL comments are skipped while parsing, so, some fragments of input might be skipped. For example, `'Europe -- Amsterdam'`. In this case, `-- Amsterdam` is excluded from the input.
2. Special chars in string values must be escaped, for instance `'E\'Twaun Moore'`

### Does this PR introduce _any_ user-facing change?
No since the parameterized SQL feature https://github.com/apache/spark/pull/38864 hasn't been released yet.

### How was this patch tested?
By running the affected tests:
```
$ build/sbt "test:testOnly *ParametersSuite"
$ python/run-tests --parallelism=1 --testnames 'pyspark.sql.tests.connect.test_connect_basic SparkConnectBasicTests.test_sql_with_args'
$ python/run-tests --parallelism=1 --testnames 'pyspark.sql.session SparkSession.sql'
```

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -234,19 +236,24 @@ class SparkSession private[sql] (
...
    * @param args
-   *   A map of parameter names to string values that are parsed as SQL literal expressions. For
-   *   example, map keys: "rank", "name", "birthdate"; map values: "1", "'Steven'",
-   *   "DATE'2023-03-21'". The fragments of string values belonged to SQL comments are skipped
-   *   while parsing.
+   *   A map of parameter names to Java/Scala objects that can be converted to SQL literal
+   *   expressions. See <a href="https://spark.apache.org/docs/latest/sql-ref-datatypes.html">
+   *   Supported Data Types</a> for supported value types in Scala/Java. For example, map keys:
+   *   "rank", "name", "birthdate"; map values: 1, "Steven", LocalDate.of(2023, 4, 2). Map value
+   *   can be a `Column` of literal expression, in that case it is taken as is.
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:243]:
```suggestion
   *   can be also a `Column` of literal expression, in that case it is taken as is.
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -215,15 +216,16 @@ class SparkSession private[sql] (
...
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:223]:
```suggestion
   *   can also be a `Column` of literal expression, in that case it is taken as is.
```

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -215,15 +216,16 @@ class SparkSession private[sql] (
    * @param sqlText
    *   A SQL statement with named parameters to execute.
    * @param args
-   *   A map of parameter names to string values that are parsed as SQL literal expressions. For
-   *   example, map keys: "rank", "name", "birthdate"; map values: "1", "'Steven'",
-   *   "DATE'2023-03-21'". The fragments of string values belonged to SQL comments are skipped
-   *   while parsing.
+   *   A map of parameter names to Java/Scala objects that can be converted to SQL literal
+   *   expressions. See <a href="https://spark.apache.org/docs/latest/sql-ref-datatypes.html">
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:220]: Here is the PR https://github.com/apache/spark/pull/40644

**Review by entong — APPROVED** [review]

**Review by srielau — APPROVED** [review]

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 4]
```diff
@@ -53,11 +53,8 @@ message SqlCommand {
   // (Required) SQL Query.
   string sql = 1;
 
-  // (Optional) A map of parameter names to string values that are parsed as
-  // SQL literal expressions. For example, map keys: "rank", "name", "birthdate";
-  // map values: "1", "'Steven'", "DATE'2023-03-21'". The fragments of string values
-  // belonged to SQL comments are skipped while parsing.
-  map<string, string> args = 2;
+  // (Optional) A map of parameter names to literal expressions.
+  map<string, Expression.Literal> args = 2;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:57]: cc @grundprinzip are these protocol changes ok?

Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:57]: No, this is an incomatible change.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 5]
```diff
@@ -108,11 +108,8 @@ message SQL {
   // (Required) The SQL query.
   string query = 1;
 
-  // (Optional) A map of parameter names to string values that are parsed as
-  // SQL literal expressions. For example, map keys: "rank", "name", "birthdate";
-  // map values: "1", "'Steven'", "DATE'2023-03-21'". The fragments of string values
-  // belonged to SQL comments are skipped while parsing.
-  map<string, string> args = 2;
+  // (Optional) A map of parameter names to literal expressions.
+  map<string, Expression.Literal> args = 2;
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:112]: Same

**Review by grundprinzip — CHANGES_REQUESTED** [review]
The Spark Connect changes are breaking.

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 4 above)
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:57]: would be valid:

```suggestion
  map<string, string> args = 2;
  map<string, Expression.Literal> expr_args = 3;
```

so might be

```suggestion
  oneof args {
    map<string, string> args = 2;
    map<string, Expression.Literal> expr_args = 3;
  }
```

Please run 

```
buf breaking --against "https://github.com/apache/spark/archive/master.zip#strip_components=1,subdir=connector/connect/common/src/main"
```

**Comment by cloud-fan** [comment]
@grundprinzip Spark Connect is not released yet, I think we can still change it? This PR should go to Spark 3.4. cc @xinrong-meng

**Comment by grundprinzip** [comment]
If it's guaranteed to be included in 3.4 it's not a breaking change.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

**Comment by cloud-fan** [comment]
It has conflicts with 3.4, @MaxGekk can you create a backport PR? Thanks!

**Comment by MaxGekk** [comment]
@cloud-fan I am working on the backport ...

**Comment by MaxGekk** [comment]
Here is the backport to `branch-3.4`: https://github.com/apache/spark/pull/40666

**Pull request closed without merging**